### PR TITLE
Strict mode: rename `config.lint` to `config.strict`, crash early on module or template error. Add `MULTIQC_STRICT=1`

### PR DESCRIPTION
**Pull request by vladsavelyev**
Rebrand the "lint mode" into the "strict mode": 
* Rename `--lint`/`config.lint` to `--strict`/`config.strict`
* Make MultiQC additionally quickly crash whenever a module or a template raised an unexpected exception
* Allow enabling the mode with `MULTIQC_STRICT=1`

The `--lint`/`config.lint` flag is still supported for back-compatibility.

### CODE REVIEW

**Comment by vladsavelyev** [comment]
@multiqc-bot changelog

File: multiqc/utils/lint_helpers.py
Comment (ewels) [inline multiqc/utils/lint_helpers.py:1]: Maybe rename this file to `strict_helpers.py`?

File: multiqc/multiqc.py [hunk 1]
```diff
@@ -233,7 +233,17 @@
     is_flag=True,
     help="Use only interactive plots [i](in-browser Javascript)[/]",
 )
-@click.option("--lint", "lint", is_flag=True, help="Use strict linting (validation) to help code development")
+@click.option(
+    "--strict",
+    "--lint",
```
Comment (ewels) [inline multiqc/multiqc.py:238]: I think have this as a separate `@click.option`, and merge in the Python code of the function. Then the CLI option can be marked as `hidden` with Click and not show up in the CLI help. Can also throw a warning log message in the Python of the main def saying that it's deprecated.

Comment (vladsavelyev) [reply multiqc/multiqc.py:238]: Yes, good idea!

File: multiqc/multiqc.py [hunk 2]
```diff
@@ -233,7 +233,17 @@
...
 )
-@click.option("--lint", "lint", is_flag=True, help="Use strict linting (validation) to help code development")
+@click.option(
+    "--strict",
+    "strict",
+    is_flag=True,
+    help="Strict mode. MultiQC will stop whenever a module or a template raised an "
+    "unexpected exception. The module code will be additionally validated "
+    "(linted) to help code development, and any validation issues will be printed "
+    "to the console in the end of the run, and also cause MultiQC to exist with "
```
Comment (ewels) [inline multiqc/multiqc.py:243]:
```suggestion
    "to the console in the end of the run, and also cause MultiQC to exit with "
```

File: multiqc/plots/linegraph.py [hunk 3]
```diff
@@ -222,25 +222,31 @@ def plot(data, pconfig=None):
             d.setdefault("color", scale.get_colour(di, lighten=1))
 
     # Make a plot - template custom, or interactive or flat
-    try:
-        return get_template_mod().linegraph(plotdata, pconfig)
-    except (AttributeError, TypeError):
```
Comment (ewels) [inline multiqc/plots/linegraph.py:227]: Huh, we just defaulted to flat plots without any log or anything? 😅

File: multiqc/plots/linegraph.py [hunk 4]
```diff
@@ -222,25 +222,31 @@ def plot(data, pconfig=None):
...
+                # debugging of modules
+                raise
+    if config.plots_force_flat or (
+        not config.plots_force_interactive and plotdata and len(plotdata[0]) > config.plots_flat_numseries
+    ):
+        try:
+            report.num_mpl_plots += 1
+            return matplotlib_linegraph(plotdata, pconfig)
+        except Exception as e:
+            logger.error("############### Error making MatPlotLib figure! Falling back to HighCharts.")
```
Comment (ewels) [inline multiqc/plots/linegraph.py:241]: Raise if strict here as well?

File: multiqc/plots/bargraph.py [hunk 5]
```diff
@@ -192,29 +192,35 @@ def plot(data, cats=None, pconfig=None):
...
+    else:
+        # Use MatPlotLib to generate static plots if requested
+        if config.export_plots:
             try:
-                report.num_mpl_plots += 1
-                return matplotlib_bargraph(plotdata, plotsamples, pconfig)
+                matplotlib_bargraph(plotdata, plotsamples, pconfig)
             except Exception as e:
-                logger.error("############### Error making MatPlotLib figure! Falling back to HighCharts.")
+                logger.error("############### Error making MatPlotLib figure! Plot not exported.")
```
Comment (ewels) [inline multiqc/plots/bargraph.py:220]: Raise if strict?

**Review by ewels — APPROVED** [review]
On my phone and the diff is a bit funky, but I think it looks good 👍🏻

**Review by ewels — CHANGES_REQUESTED** [review]
Needs docs

File: multiqc/plots/linegraph.py (hunk 3 above)
Comment (vladsavelyev) [reply multiqc/plots/linegraph.py:227]: Doesn't seem so? 🤔

**Comment by vladsavelyev** [comment]
> Needs docs

I modified/extended the existing docs about the lint mode. Should we add a completely separate section about that? It's just a small addition to the previously existing lint mode, so probably adding a few sentences to the existing sections is enough.

File: multiqc/multiqc.py [hunk 6]
```diff
@@ -233,7 +233,17 @@
...
-@click.option("--lint", "lint", is_flag=True, help="Use strict linting (validation) to help code development")
+@click.option(
+    "--strict",
+    "strict",
+    is_flag=True,
+    help="Strict mode. MultiQC will stop whenever a module or a template raised an "
+    "unexpected exception. The module code will be additionally validated "
+    "(linted) to help code development, and any validation issues will be printed "
+    "to the console in the end of the run, and also cause MultiQC to exit with "
+    "a non-0 code",
```
Comment (ewels) [inline multiqc/multiqc.py:244]: This help text is much longer than all other options. It'd be good to trim to about a 1/3rd so it's comparable to other options. Can always include a hyperlink to the docs if we want!

```suggestion
    help="Don't catch exceptions, run additional code checks to help development.",
```

**Review by ewels — APPROVED** [review]
Minor comment with CLI help text, otherwise LGTM to merge 👍🏻

**Comment by ewels** [comment]
Docs look good - I wondered about adding something to "contributing" but I think it's good for now. That's more part of a larger docs restructuring that I think we should tackle at some point.

**Pull request merged**